File: app/src/main/java/com/sktpj/td2048/GameRules.kt
package com.sktpj.td2048

import kotlin.random.Random

internal data class LineMergeResult(
    val line: List<Int>,
    val createdValues: List<Int>,
)

internal data class BoardMoveResult(
    val board: List<Int>,
    val moved: Boolean,
    val createdValues: List<Int>,
)

internal object GameRules {
    const val GRID_SIZE = 4
    const val CELL_COUNT = GRID_SIZE * GRID_SIZE

    fun initialBoard(random: Random): List<Int> =
        spawnRandomTile(spawnRandomTile(List(CELL_COUNT) { 0 }, random), random)

    fun slideAndMergeLine(line: List<Int>): LineMergeResult {
        require(line.size == GRID_SIZE)
        val compact = line.filter { it != 0 }
        val merged = mutableListOf<Int>()
        val created = mutableListOf<Int>()
        var index = 0
        while (index < compact.size) {
            if (index + 1 < compact.size && compact[index] == compact[index + 1]) {
                val value = compact[index] * 2
                merged += value
                created += value
                index += 2
            } else {
                merged += compact[index]
                index += 1
            }
        }
        while (merged.size < GRID_SIZE) merged += 0
        return LineMergeResult(merged, created)
    }

    fun moveWithoutSpawn(board: List<Int>, direction: Direction): BoardMoveResult {
        require(board.size == CELL_COUNT)
        val next = MutableList(CELL_COUNT) { 0 }
        val created = mutableListOf<Int>()
        var moved = false

        for (outer in 0 until GRID_SIZE) {
            val indices = buildList {
                for (inner in 0 until GRID_SIZE) {
                    add(
                        when (direction) {
                            Direction.LEFT -> cellIndex(outer, inner)
                            Direction.RIGHT -> cellIndex(outer, GRID_SIZE - 1 - inner)
                            Direction.UP -> cellIndex(inner, outer)
                            Direction.DOWN -> cellIndex(GRID_SIZE - 1 - inner, outer)
                        },
                    )
                }
            }
            val result = slideAndMergeLine(indices.map(board::get))
            created += result.createdValues
            result.line.forEachIndexed { lineIndex, value ->
                val boardIndex = indices[lineIndex]
                next[boardIndex] = value
                if (value != board[boardIndex]) moved = true
            }
        }
        return BoardMoveResult(next, moved, created)
    }

    fun spawnRandomTile(board: List<Int>, random: Random): List<Int> {
        val empties = board.indices.filter { board[it] == 0 }
        if (empties.isEmpty()) return board
        val target = empties[random.nextInt(empties.size)]
        return board.toMutableList().also {
            it[target] = if (random.nextFloat() < 0.9f) 2 else 4
        }
    }

    fun canMove(board: List<Int>): Boolean {
        if (board.any { it == 0 }) return true
        for (index in board.indices) {
            val row = index / GRID_SIZE
            val col = index % GRID_SIZE
            val value = board[index]
            if (col < GRID_SIZE - 1 && board[cellIndex(row, col + 1)] == value) return true
            if (row < GRID_SIZE - 1 && board[cellIndex(row + 1, col)] == value) return true
        }
        return false
    }

    fun cellIndex(row: Int, col: Int): Int = row * GRID_SIZE + col
    fun rowOf(cellIndex: Int): Int = cellIndex / GRID_SIZE
    fun colOf(cellIndex: Int): Int = cellIndex % GRID_SIZE
}

File: docs/src/game_rules.js
export const GRID_SIZE = 4;
export const CELL_COUNT = GRID_SIZE * GRID_SIZE;

export function cellIndex(row, col) { return row * GRID_SIZE + col; }
export function rowOf(index) { return Math.floor(index / GRID_SIZE); }
export function colOf(index) { return index % GRID_SIZE; }

export function slideAndMergeLine(line) {
  if (line.length !== GRID_SIZE) throw new Error("line must have 4 cells");
  const compact = line.filter((value) => value !== 0);
  const merged = [];
  const createdValues = [];
  for (let i = 0; i < compact.length;) {
    if (i + 1 < compact.length && compact[i] === compact[i + 1]) {
      const value = compact[i] * 2;
      merged.push(value);
      createdValues.push(value);
      i += 2;
    } else {
      merged.push(compact[i]);
      i += 1;
    }
  }
  while (merged.length < GRID_SIZE) merged.push(0);
  return { line: merged, createdValues };
}

export function moveWithoutSpawn(board, direction) {
  if (board.length !== CELL_COUNT) throw new Error("board must have 16 cells");
  const next = Array(CELL_COUNT).fill(0);
  const createdValues = [];
  let moved = false;

  for (let outer = 0; outer < GRID_SIZE; outer += 1) {
    const indices = [];
    for (let inner = 0; inner < GRID_SIZE; inner += 1) {
      if (direction === "LEFT") indices.push(cellIndex(outer, inner));
      else if (direction === "RIGHT") indices.push(cellIndex(outer, GRID_SIZE - 1 - inner));
      else if (direction === "UP") indices.push(cellIndex(inner, outer));
      else if (direction === "DOWN") indices.push(cellIndex(GRID_SIZE - 1 - inner, outer));
      else throw new Error(`unknown direction: ${direction}`);
    }
    const result = slideAndMergeLine(indices.map((index) => board[index]));
    createdValues.push(...result.createdValues);
    result.line.forEach((value, lineIndex) => {
      const index = indices[lineIndex];
      next[index] = value;
      if (value !== board[index]) moved = true;
    });
  }
  return { board: next, moved, createdValues };
}

export function spawnRandomTile(board, random = Math.random) {
  const empties = board.map((value, index) => value === 0 ? index : -1).filter((index) => index >= 0);
  if (empties.length === 0) return board.slice();
  const target = empties[Math.floor(random() * empties.length)];
  const next = board.slice();
  next[target] = random() < 0.9 ? 2 : 4;
  return next;
}

export function initialBoard(random = Math.random) {
  return spawnRandomTile(spawnRandomTile(Array(CELL_COUNT).fill(0), random), random);
}

export function canMove(board) {
  if (board.some((value) => value === 0)) return true;
  for (let index = 0; index < board.length; index += 1) {
    const row = rowOf(index);
    const col = colOf(index);
    const value = board[index];
    if (col < GRID_SIZE - 1 && board[cellIndex(row, col + 1)] === value) return true;
    if (row < GRID_SIZE - 1 && board[cellIndex(row + 1, col)] === value) return true;
  }
  return false;
}
